docs/Toy/toy/parser.py

```diff
@@ -0,0 +1,437 @@
+from pathlib import Path
+from typing import List, NoReturn, TypeVar, cast
+
+from .lexer import (OperatorToken, Token, tokenize, NumberToken, EOFToken,
+                    IdentifierToken)
+from .toy_ast import (BinaryExprAST, ExprAST, NumberExprAST, FunctionAST,
+                      ModuleAST, LiteralExprAST, VarDeclExprAST, VarType,
+                      VariableExprAST, ReturnExprAST, PrintExprAST,
+                      CallExprAST, PrototypeAST)
+
+
+class ParseError(Exception):
+
+    def __init__(self,
+                 token: Token,
+                 expected: str | type[Token],
+                 context: str = '',
+                 line: str = ''):
+        loc = token.loc
+
+        message = f'Parse error ({loc.line}, {loc.col}): expected '
+        if isinstance(expected, str):
+            message += expected
+        else:
+            message += expected.name()
+
+        if len(context):
+            message += ' ' + context
+
+        message += f" but has Token '{token.text}'\n"
+
+        if len(line):
+            message += line + '\n'
+
+        super().__init__(message)
+
+    pass
+
+
+TokenT = TypeVar('TokenT', bound=Token)
+
+
+class Parser:
+    file: Path
+    program: str
+    tokens: List[Token]
+    pos: int
+
+    def __init__(self, file: Path, program: str):
+        self.file = file
+        self.program = program
+        self.tokens = tokenize(file, program)
+        self.pos = 0
+
+    def getToken(self):
+        '''Returns current token in parser'''
+        return self.tokens[self.pos]
+
+    def getTokenPrecedence(self) -> int:
+        '''Returns precedence if the current token is a binary operation, -1 otherwise'''
```

Comment:
Don;t we want """ quotes?

Reply:
Do we? I guess I'm used to single quotes, but consistency is best. It looks like I'm the only one who's added triple single quotes in the whole repo, I'll undo that in this one, and a followup for the code that I've already checked in